### PR TITLE
Update global jdas enkf diag job with COMIN/COMOUT for COM prefix

### PR DESCRIPTION
**Pull request by mingshichen-noaa**
# Description
 
NCO has requested that each COM variable specify whether it is an input or an output. This completes that process for the global jdas enkf diagnostics job.

Refs https://github.com/NOAA-EMC/global-workflow/issues/2451

# Type of change

- Maintenance (code refactor, clean-up, new CI test, etc.)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?
 
- Clone and build on RDHPCS
- Cycled tests on Hercules
- Forecast-only tests on Hercules
 
# Checklist
- [x] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] I have made corresponding changes to the documentation if necessary

### CODE REVIEW

File: jobs/JGDAS_ENKF_DIAG
Comment (KateFriedman-NOAA) [inline jobs/JGDAS_ENKF_DIAG]: This should be `COMIN_ATMOS_ANALYSIS_DET_PREV` since it's a location that already exists and is where some input files will come from. Update the declare to make it a `COMIN_*` path and update the variable in the rest of the scripts as well.

File: jobs/JGDAS_ENKF_DIAG
Comment (KateFriedman-NOAA) [inline jobs/JGDAS_ENKF_DIAG]: Same comment, these should be `COMIN_ATMOS_ANALYSIS_DET_PREV`.

File: jobs/JGDAS_ENKF_DIAG
Comment (KateFriedman-NOAA) [inline jobs/JGDAS_ENKF_DIAG]: Same comment, should be a `COMIN` since it's where inputs will come from. Change to `COMIN_ATMOS_HISTORY_PREV`.

File: jobs/JGDAS_ENKF_DIAG
Comment (KateFriedman-NOAA) [inline jobs/JGDAS_ENKF_DIAG]: Same comment, should be a `COMIN`, change to `COMIN_ATMOS_HISTORY_PREV`. In general, all previous cycle `COM`s (e.g. anything with `PREV` in the name) are `COMIN`s.

**Review by KateFriedman-NOAA — CHANGES_REQUESTED** [review]
Changes look ok for the most part but some `COMOUT`s needs to be changed to `COMIN`s.

**Comment by mingshichen-noaa** [comment]
@KateFriedman-NOAA 
Thank you very much for your comments. Based on your suggestions. I have changed "COMOUT_ATMOS_ANALYSIS_DET_PREV" into "COMIN_ATMOS_ANALYSIS_DET_PREV".

**Review by KateFriedman-NOAA — DISMISSED** [review]
Requested updates alongside other updates look good, thanks @mingshichen-noaa !

**Review by WalterKolczynski-NOAA — DISMISSED** [review]

**Comment by WalterKolczynski-NOAA** [comment]
JGDAS_ATMOS_ANALYSIS_DIAG also uses the exglobal_diag.sh script, so they have to be updated at the same time.

```
+ JGDAS_ATMOS_ANALYSIS_DIAG[30]: declare_from_tmpl -rx COM_ATMOS_ANALYSIS
+ bash_utils.sh[35]: [[ NO == \N\O ]]
+ bash_utils.sh[35]: set +x
declare_from_tmpl :: COM_ATMOS_ANALYSIS=/lfs/h2/emc/global/noscrub/globalworkflow.ci/GFS_CI_ROOT/PR/2959/RUNTESTS/COMROOT/C96C48_hybatmDA_c55c8630/gdas.20211221/00//analysis/atmos
+ JGDAS_ATMOS_ANALYSIS_DIAG[31]: mkdir -m 775 -p /lfs/h2/emc/global/noscrub/globalworkflow.ci/GFS_CI_ROOT/PR/2959/RUNTESTS/COMROOT/C96C48_hybatmDA_c55c8630/gdas.20211221/00//analysis/atmos
+ JGDAS_ATMOS_ANALYSIS_DIAG[35]: /lfs/h2/emc/global/noscrub/globalworkflow.ci/GFS_CI_ROOT/PR/2959/global-workflow/scripts/exglobal_diag.sh
Begin exglobal_diag.sh at Fri Sep 27 06:36:46 UTC 2024
++ exglobal_diag.sh[25]: pwd
+ exglobal_diag.sh[25]: pwd=/lfs/h2/emc/stmp/terry.mcguinness/RUNDIRS/C96C48_hybatmDA_c55c8630/gdas.2021122100/analdiag.246715
+ exglobal_diag.sh[28]: CDATE=2021122100
+ exglobal_diag.sh[29]: GDUMP=gdas
+ exglobal_diag.sh[32]: export 'CHGRP_CMD=chgrp rstprod'
+ exglobal_diag.sh[32]: CHGRP_CMD='chgrp rstprod'
+ exglobal_diag.sh[33]: export NCLEN=/lfs/h2/emc/global/noscrub/globalworkflow.ci/GFS_CI_ROOT/PR/2959/global-workflow/ush/getncdimlen
+ exglobal_diag.sh[33]: NCLEN=/lfs/h2/emc/global/noscrub/globalworkflow.ci/GFS_CI_ROOT/PR/2959/global-workflow/ush/getncdimlen
+ exglobal_diag.sh[34]: export CATEXEC=/apps/ops/prod/libs/intel/19.1.3.304/cray-mpich/8.1.4/ncdiag/1.0.0/bin/ncdiag_cat_serial.x
+ exglobal_diag.sh[34]: CATEXEC=/apps/ops/prod/libs/intel/19.1.3.304/cray-mpich/8.1.4/ncdiag/1.0.0/bin/ncdiag_cat_serial.x
+ exglobal_diag.sh[35]: COMPRESS=gzip
+ exglobal_diag.sh[36]: UNCOMPRESS=gunzip
+ exglobal_diag.sh[37]: APRUNCFP='mpiexec -l -np $ncmd --cpu-bind verbose,core cfp'
+ exglobal_diag.sh[40]: netcdf_diag=.true.
+ exglobal_diag.sh[41]: binary_diag=.false.
+ exglobal_diag.sh[44]: RUN=gdas
+ exglobal_diag.sh[45]: SENDECF=NO
+ exglobal_diag.sh[46]: SENDDBN=NO
+ exglobal_diag.sh[51]: export APREFIX=gdas.t00z.
+ exglobal_diag.sh[51]: APREFIX=gdas.t00z.
/lfs/h2/emc/global/noscrub/globalworkflow.ci/GFS_CI_ROOT/PR/2959/global-workflow/scripts/exglobal_diag.sh: line 52: COMOUT_ATMOS_ANALYSIS: unbound variable
```

**Comment by mingshichen-noaa** [comment]
@WalterKolczynski-NOAA: 
Your comment: "JGDAS_ATMOS_ANALYSIS_DIAG also uses the exglobal_diag.sh script, so they have to be updated at the same time."
In fact, I have updated both GDAS_ATMOS_ANALYSIS_DIAG and  exglobal_diag.sh script in this PR. 
@RaghuReddy-NOAA:: 
Today you told me the jobs (JGDAS_ENKF_*) are relevant each other, this means I should update all jobs (JGDAS_ENKF_*) and associated bash and python scripts in one PR at the same time. Am I right?

The jobs (JGDAS_ENKF_*) include JGDAS_ENKF_DIAG,  JGDAS_ENKF_ECEN , JGDAS_ENKF_POST,  JGDAS_ENKF_SELECT_OBS,  JGDAS_ENKF_SFC  , JGDAS_ENKF_SNOW_RECENTER,  JGDAS_ENKF_UPDATE

**Comment by WalterKolczynski-NOAA** [comment]
> @WalterKolczynski-NOAA: Your comment: "JGDAS_ATMOS_ANALYSIS_DIAG also uses the exglobal_diag.sh script, so they have to be updated at the same time." In fact, I have updated both GDAS_ATMOS_ANALYSIS_DIAG and exglobal_diag.sh script in this PR. @RaghuReddy-NOAA:: Today you told me the jobs (JGDAS_ENKF__) are relevant each other, this means I should update all jobs (JGDAS_ENKF__) and associated bash and python scripts in one PR at the same time. Am I right?
> 
> The jobs (JGDAS_ENKF_*) include JGDAS_ENKF_DIAG, JGDAS_ENKF_ECEN , JGDAS_ENKF_POST, JGDAS_ENKF_SELECT_OBS, JGDAS_ENKF_SFC , JGDAS_ENKF_SNOW_RECENTER, JGDAS_ENKF_UPDATE

No, you only need to do the ones that are calling the script you modify here (`exglobal_diag.sh`), that's just two j-job scripts:
```
>grep exglobal_diag.sh jobs/*
jobs/JGDAS_ATMOS_ANALYSIS_DIAG:${ANALDIAGSH:-${SCRgfs}/exglobal_diag.sh}
jobs/JGDAS_ENKF_DIAG:${ANALDIAGSH:-${SCRgfs}/exglobal_diag.sh}
```

**Comment by RaghuReddy-NOAA** [comment]
@WalterKolczynski-NOAA  and @mingshichen-noaa  I see I am tagged in the two messages above, I am wondering if it was intended for Rahul?

**Comment by mingshichen-noaa** [comment]
@RaghuReddy-NOAA 
I am sorry that the previous messages to you and Walter at  a window  have tagged you. Now let me explain my question. below
Because The jobs (JGDAS_ENKF_*: including JGDAS_ENKF_DIAG, JGDAS_ENKF_ECEN , JGDAS_ENKF_POST, JGDAS_ENKF_SELECT_OBS, JGDAS_ENKF_SFC , JGDAS_ENKF_SNOW_RECENTER, JGDAS_ENKF_UPDATE) are associate with JDAS EnKF,  can I create one branch to change COM_ into COMIN/COMOUT_ instead of job by job?

**Comment by DavidHuber-NOAA** [comment]
@mingshichen-noaa Is this PR ready to be reviewed/tested again?

**Review by WalterKolczynski-NOAA — DISMISSED** [review]
Looks correct, but CI will need to wait until #2928 is merged and then develop merged into this PR, as that PR contains fixes needed for one of the CI tests.

**Comment by WalterKolczynski-NOAA** [comment]
@mingshichen-noaa Before we run this through automated CI, did this pass a test using the C96_atm3DVar case?

**Comment by mingshichen-noaa** [comment]
@WalterKolczynski-NOAA 
Yesterday I tested PR #2959 in 3 forecast-only cases (C48_ATM  C48_S2SW  C48_S2SWA_gefs) and 3 cycled cases(  C96_atm3DVar   C96C48_hybatmDA  C96C48_ufs_hybatmDA) in hercules environment.
Today I tested C96_atm3DVar in Orion environment
After I did " rocotostat -d C96_atm3DVar.db -w C96_atm3DVar.xml -v 10" , head lines of results showed as follows:
       CYCLE                    TASK                       JOBID               STATE         EXIT STATUS     TRIES      DURATION
================================================================================================================================
202112201800           gdas_stage_ic    druby://130.18.14.111:34781          SUBMITTING                   -         0           0.0
202112201800          gdas_fcst_seg0                           -                   -                   -         -             -
202112201800    gdas_atmos_prod_f000                           -                   -                   -         -             -
202112201800    gdas_atmos_prod_f003                           -                   -                   -         -             -
202112201800    gdas_atmos_prod_f006                           -                   -                   -         -             -
202112201800    gdas_atmos_prod_f009                           -                   -                   -         -             -
================================================================================================================================

**Review by CoryMartin-NOAA — DISMISSED** [review]

**Comment by WalterKolczynski-NOAA** [comment]
> @WalterKolczynski-NOAA
> Yesterday I tested PR #2959 in 3 forecast-only cases (C48_ATM C48_S2SW C48_S2SWA_gefs) and 3 cycled cases( C96_atm3DVar C96C48_hybatmDA C96C48_ufs_hybatmDA) in hercules environment.
> Today I tested C96_atm3DVar in Orion environment
> After I did " rocotostat -d C96_atm3DVar.db -w C96_atm3DVar.xml -v 10" , head lines of results showed as follows:
> CYCLE TASK JOBID STATE EXIT STATUS TRIES DURATION
> 202112201800 gdas_stage_ic druby://130.18.14.111:34781 SUBMITTING - 0 0.0
> 202112201800 gdas_fcst_seg0 - - - - -
> 202112201800 gdas_atmos_prod_f000 - - - - -
> 202112201800 gdas_atmos_prod_f003 - - - - -
> 202112201800 gdas_atmos_prod_f006 - - - - -
> 202112201800 gdas_atmos_prod_f009 - - - - -

Do you have an entry for `rocotorun` in you crontab for each experiment? Because it looks like it hasn't run since the first time.

File: jobs/JGDAS_ATMOS_ANALYSIS_DIAG
Comment (DavidHuber-NOAA) [inline jobs/JGDAS_ATMOS_ANALYSIS_DIAG]: Should be declared from the template:

```suggestion
	COMOUT_ATMOS_ANALYSIS:COM_ATMOS_ANALYSIS_TMPL
```

**Review by DavidHuber-NOAA — APPROVED** [review]

**Review by KateFriedman-NOAA — APPROVED** [review]
Approve pending successful completion of CI tests. Thanks @mingshichen-noaa !

**Comment by WalterKolczynski-NOAA** [comment]
Running in `/lfs/h2/emc/ptmp/walter.kolczynski/PR/PR_2959`

**Pull request merged**
